### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by Luzaks**
In this branch we create:
* a Posts, Users and Sessions controllers
* views for the signup page, login page, user info, index all posts and create all posts
* two Models, Post, and User
* methods for remembering a user session
* methods for secret post's author's name.

### CODE REVIEW

File: app/helpers/sessions_helper.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def log_out
+    session.delete(:user_id)
+    @current_user = nil
+  end
+end
```
Comment (Redvanisation) [inline app/helpers/sessions_helper.rb]: Please move those methods to the **ApplicationController**.

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

### Hello team  🌟

**Very nice work on this project, however, the following issues need to be fixed before moving on ⬇️** 

### ApplicationController
* [x] It uses the #current_user method to retrive user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).
* [x] It uses the #current_user=(user) method to set user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).
* [x] It has a sign_in method ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).
* [x] It has a sign_out method ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/session_delete_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).

* [x] There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers, check [this article](http://codefol.io/posts/Where-Do-I-Put-My-Code) for reference.

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    @logged_in = logged_in?
+  end
+
+  def new
+    @post = Post.new
+  end
+
+  def create
+    @post = Post.new(post_params)
+    @post.user_id = current_user.id
```
Comment (efrapp) [inline app/controllers/posts_controller.rb]: [`Optional`] Here you can assign the `current_user` to the `@post.user` directly. No need to use the ids.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+Rails.application.routes.draw do
+  root 'posts#index'
+  get    '/signup',  to: 'users#new'
+  get    '/login',   to: 'sessions#new'
+  post   '/login',   to: 'sessions#create'
+  delete '/logout',  to: 'sessions#destroy'
+  post 'new', to: 'posts#create'
+  resources :users
+  resources :posts, :only => [:new, :create, :index]
```
Comment (efrapp) [inline config/routes.rb]: If you already have a `resources :posts` no need to use the `post 'new' ...` route.

**Review by efrapp — CHANGES_REQUESTED** [review]
## Hi @Luzaks and @mcastorena0316 👋 👋
You are doing an excellent job in this project, I like your interface and your code is clear. Please, check below to check my comments about the project.

### From the last review
- [X] It uses the #current_user method to retrieve the user
- [x] It uses the `current_user=(user)` method to set user: I saw in your `log` folder that you tried to use this method. This method is just to assign the value passed as a parameter (the user) to the current_user instance variable. **Please, add this method**.
- [X] It has a `sign_in` method
- [X] It has a `sign_out` method
- [X] There should be NO application logic in the views nor helpers

### Suggestions (optional)
- Please, always use a `.gitignore` file: you are tracking the `node_modules` folder in your project. This folder weighs 69,7 MB making your project too heavy.  Also, that folder is not needed because it will be recreated locally with the `package.json` file. When you create a new Rails project (using the command) you will be provided with a `.gitignore` file by default so you can use that to start.
- Always try to show to the user what is happening in case of errors. For instance, you don't display any error message if the user tries to signup with the fields empty
- Try always to make accessible the different parts of the page, using a navbar for example

Ok team, there is just a task missing, create the `current_use=(user)` method and please check the suggestions I made above and on the code.

Send a new code review when you are done with the change.

Have a great day!

**Status:** `Changes required`

File: app/views/posts/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+<h1>Comment</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  <% @posts.each do |p|%>
+    <ul>
+    <li> Comment: <%= p.body %>,  </li>
+    <% if @logged_in %>
+     Written by:    <%= p.user.name %>,
+    <% end %>
```
Comment (ebukaume) [inline app/views/posts/index.html.erb:11]: [OPTIONAL]
You should have made `logged_in?' a helper so that way you just use it directly here in the view.

**Review by ebukaume — APPROVED** [review]
## Approved :champagne: 

Hi @Luzaks  and @mcastorena0316 :wave: 

You guys have done well. **You can go ahead and merge this branch.**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Pull request merged**